kart/spatial_tree.py

```diff
@@ -0,0 +1,413 @@
+        return transform, desc
+
+
+class SpatialTreeTables(TableSet):
+    """Spatial Tree spec tables - see http://www.geopackage.org/spec/#table_definition_sql"""
+
+    def __init__(self):
+        super().__init__()
+
+        self.blobs = Table(
+            "blobs",
+            self.sqlalchemy_metadata,
+            Column("rowid", Integer, nullable=False, primary_key=True),
+            Column("blob_id", BLOB, nullable=False, unique=True),
+            sqlite_autoincrement=True,
+        )
+
+        self.blob_cells = Table(
+            "blob_cells",
+            self.sqlalchemy_metadata,
+            Column(
+                "blob_rowid",
+                Integer,
+                ForeignKey("blobs.rowid"),
+                nullable=False,
+                primary_key=True,
+            ),
+            Column(
+                "cell_token",
+                Text,
+                nullable=False,
+                primary_key=True,
+            ),
+        )
+
+
+SpatialTreeTables.copy_tables_to_class()
+
+
+def drop_tables(sess):
+    sess.execute("DROP TABLE IF EXISTS blob_cells;")
+    sess.execute("DROP TABLE IF EXISTS blobs;")
+
+
+def iter_feature_oids(repo, commit_spec):
+    cmd = _revlist_command(repo) + commit_spec
+    try:
+        p = subprocess.Popen(
+            cmd,
+            stdout=subprocess.PIPE,
+            encoding="utf8",
+            env=tool_environment(),
+        )
+        yield from _parse_revlist_output(p.stdout, r"feature/.+")
+    except subprocess.CalledProcessError as e:
+        raise SubprocessError(
+            f"There was a problem with git rev-list: {e}", called_process_error=e
+        )
+
+
+def update_spatial_tree(
+    repo, commit_spec, crs_helper, verbosity=1, clear_existing=False
+):
+    import pywraps2 as s2
+
+    feature_oid_iter = iter_feature_oids(repo, commit_spec)
+
+    s2_coverer = s2.S2RegionCoverer()
+    s2_coverer.set_max_cells(S2_MAX_CELLS_INDEX)
+    s2_coverer.set_max_level(S2_MAX_LEVEL)
+
+    progress_every = None
+    if verbosity >= 1:
+        progress_every = max(100, 100_000 // (10 ** (verbosity - 1)))
+
+    db_path = repo.gitdir_file(KartRepoFiles.S2_INDEX)
+    engine = sqlite_engine(db_path)
+    with sessionmaker(bind=engine)() as sess:
+        if clear_existing:
+            drop_tables(sess)
+
+        SpatialTreeTables.create_all(sess)
+
+    t0 = time.monotonic()
+    click.echo(f"Indexing {' '.join(commit_spec)} ...")
+
+    # Using sqlite directly here instead of sqlalchemy is about 10x faster.
+    # Possibly due to huge number of unbatched queries.
+    # TODO - investigate further.
+    db = sqlite.connect(f"file:{db_path}", uri=True)
+    with db:
+        dbcur = db.cursor()
+
+        for i, (ds_path, feature_oid) in enumerate(feature_oid_iter):
+            if i and progress_every and i % progress_every == 0:
+                click.echo(f"  {i:,d} features... @{time.monotonic()-t0:.1f}s")
```

Comment:
I think this should go to stderr

Reply:
Not done - it's not an error and `kart import` also writes this to stdout. If this needs to change, will change them both in a separate commit.

Reply:
separate commit is fine, but I think it should go to stderr. it's debatable but e.g. in `kart diff` it's more obvious the diff goes to stdout and any kind of progress or logging must go to stderr to avoid messing up the diff. Here there's not really any stdout but to be consistent logging/progress should still go to stderr